keras_cv/losses/dice.py

```diff
@@ -0,0 +1,167 @@
+                in the combined score. If `beta < 1`, precisoin will doninate;
+                if `beta > 1`, recall will dominate. Default to `1`.
+            from_logits: Whether `y_pred` is expected to be a logits tensor. By
+                default, we assume that `y_pred` encodes a probability distribution.
+                Default to `False`.
+            class_ids: An interger or a lost of intergers within `range(num_classes)`
+                to evaluate the loss. If it's `None`, all classes will beu used to
+                calculate the loss. Default to `None`.
+            label_smoothing: Float in [0, 1]. When > 0, label values are smoothed,
+                meaning the confidence on label values are relaxed. For example, if
+                `0.1`, use `0.1 / num_classes` for non-target labels and
+                `0.9 + 0.1 / num_classes` for target labels. Default to `0.0`.
+            per_sample: If `True`, the loss will be calculated for each sample in
+                batch and then averaged. Otherwise the loss will be calculated for
+                the whole batch. Default to `False`.
+            epsilon: Small float added to dice score to avoid dividing by zero.
+                Default to `keras.backend.epsilon()` or `1e-07`.
+            name: Optional name for the instance.
+                Defaults to 'dice'.
+        """
+        super().__init__(name=name, **kwargs)
+        self.beta = beta
+        self.from_logits = from_logits
+        self.label_smoothing = label_smoothing
+        self.per_sample = per_sample
+        self.epsilon = epsilon
+
+        if class_ids is not None:
+            if isinstance(class_ids, float) or any(
+                isinstance(x, float) for x in class_ids
+            ):
+                raise ValueError(
+                    f"The indices should be int or a list of integer. Got {class_ids}"
+                )
+            elif isinstance(class_ids, int):
+                class_ids = [class_ids]
+
+        self.class_ids = class_ids
+
+    def _smooth_labels(self, y_true, y_pred, label_smoothing):
+        num_classes = tf.cast(tf.shape(y_true)[-1], y_pred.dtype)
+        return y_true * (1.0 - label_smoothing) + (label_smoothing / num_classes)
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.convert_to_tensor(y_pred)
+        y_true = tf.cast(y_true, y_pred.dtype)
+        label_smoothing = tf.convert_to_tensor(self.label_smoothing, dtype=y_pred.dtype)
+
+        if self.from_logits:
+            y_pred = tf.nn.softmax(y_pred)
+
+        if tf.cast(label_smoothing, dtype=tf.bool):
+            y_true = self._smooth_labels(y_true, y_pred, label_smoothing)
+
+        if self.class_ids is not None:
+            y_true, y_pred = gather_channels(y_true, y_pred, indices=self.class_ids)
+
+        axes = (
+            tf.constant([1, 2])
+            if keras.backend.image_data_format() == "channels_last"
+            else tf.constant([2, 3])
+        )
```

Comment:
To support `3D`  modeling, I may need to do `tf.constant([1, 2, 3]) / tf.constant([2, 3, 4])`. But not sure how to set parameters here by maintaining the keras API design. There's no 3D version of [keras-loss](https://keras.io/api/losses/) to follow guidelines. So, need some advice here.  

An example below, (not a keras style though, asaik).

```python
if modeling == '2D':
   axes = tf.constant([1, 2])
elif modeling == '3D':
   axes = tf.constant([1, 2, 3])
else:
   raise ValueError('...')
```

Reply:
Thanks, @bhack. 
I was thinking about how to introduce `generalized-dice loss` here. Seeing [this](https://github.com/tensorflow/models/blob/master/official/projects/volumetric_models/losses/segmentation_losses.py#L38-L39) implementation, I think we can introduce both params if it's ok to keras-team. 

```python
def __init__(self,
      beta,
      from_logists,
      ....
      loss_type,  # 'generalized', 'adaptive', None.
      axis= (1, 2, 3)  # (1, 2, 3) for 3D, (1, 2) for 2D.
):
```

Reply:
Added the `axis` param.